frontend/src/util/compile.ts

```diff
@@ -1,37 +1,151 @@
 // Can only use these imports as types, the actual import needs to be async
-import type { HardwareSpec, compile } from 'gdlk_wasm';
+import {
+  HardwareSpec,
+  SourceElement,
+  compile,
+  ProgramSpec,
+  Machine,
+} from 'gdlk_wasm';
+import { LangValue } from 'state/ide';
+import { isTypedArray } from './guards';
 
-export type CompiledRes = { instructions: Instruction[] } | { errors: Error[] };
+export interface CompileErrors {
+  errors: SourceElement[];
+}
+
+/**
+ * Type guard for the SourceElement type. Checks if the given value fits the
+ * criteria of a SourceElement.
+ * @param value The value to check
+ * @returns true if the value is a SourceElement, false if not
+ */
+function isSourceElement(value: object): value is SourceElement {
+  // TODO make this more robust after
+  // https://github.com/Microsoft/TypeScript/issues/21732
+  return 'text' in value && 'span' in value;
+}
+
+export interface MachineState {
```

Comment:
Could we export the struct for `Machine` and use that as a type here?

Reply:
unfortunately no, it doesn't really work if we use the `Machine` type in the react code. the issue is that react re-renders when it sees that props/context have changed. but with the `Machine` wasm type, as far as react is concerned, the value is just a pointer. So the program counter/input/output etc in the machine might change, but the pointer doesn't change so react never rerenders. we need this wrapping logic to get react to re-render after changes.